### PR TITLE
added presentation_definition endpoint

### LINKED ISSUE #2491: use presentation definition endpoint to get presentation definition from other party
The presentation definition endpoint is not defined in any standard.
We place it under `/iam/:did/presentation_definition?scope=`

### PR DESCRIPTION
**Pull request by woutslakhorst**
closes #2491 

This RP adds the presentation definition endpoint and client API. A new variable has been added to the OAuth Authorization Server Metadata: `presentation_definition_endpoint`

### CODE REVIEW

File: auth/api/iam/api.go
Comment (gerardsn) [inline auth/api/iam/api.go:263]: `scope` can contain multiple space separated scopes. This needs to be checked.

File: docs/_static/auth/iam.yaml
Comment (gerardsn) [inline docs/_static/auth/iam.yaml:150]: OAuth defines `scope` as a single space delimited string. We have the freedom to do what we want here, but in all the OAuth/OpenID defined flows the `scope` parameter is always going to be a single string. For consistency I'd do the same here.

File: docs/_static/auth/iam.yaml
Comment (gerardsn) [inline docs/_static/auth/iam.yaml:156]: copy-pasta description?

File: auth/api/iam/api.go
Comment (gerardsn) [inline auth/api/iam/api.go:263]: This generates a PD per scope while we need a PD per consent backend?

Comment (gerardsn) [reply auth/api/iam/api.go:263]: Or are we providing access per scope? (I always thought it was accept/reject entire access token request)

File: docs/_static/auth/iam.yaml
Comment (gerardsn) [inline docs/_static/auth/iam.yaml:162]: OpenID4VP flow requires a single PD for a similar request. Perhaps we should combine the different PDs in to single PD and add a `group` label per PD source. This allows selection of a single group to comply with within the combined PD using `Submission Requirements`.

**Review by gerardsn — CHANGES_REQUESTED** [review]
dealing with some of comments should probably deferred

File: auth/api/iam/api.go
Comment (reinkrul) [reply auth/api/iam/api.go:254]: Since the s2s-flow is custom, I don't see many other implementations and users of the endpoint. So validation could involve getting the definition again given the scope, and comparing it against the submitted PE?

File: auth/api/iam/client.go
Comment (reinkrul) [inline auth/api/iam/client.go:87]: This assumes very tight coupling with the calling code ("previous step"). And clients always handle input encoding, so no need to explicitly that that.

```suggestion
// PresentationDefinition retrieves the presentation definition from the presentation definition endpoint (as specified by RFC021) for the given scope.
```

File: auth/api/iam/client.go
Comment (reinkrul) [inline auth/api/iam/client.go:93]: or should they be space-separated? An OAuth2 access token (or request) always has a single scope (but it can actually consist of multiple space-separated scopes)

File: auth/api/iam/client.go
Comment (reinkrul) [inline auth/api/iam/client.go:89]: if you want a `url.URL`, you can just ask for it (string -> url.URL)?

File: auth/api/iam/client.go
Comment (reinkrul) [inline auth/api/iam/client.go:115]: better not include the data, could be binary (or very large).

File: auth/api/iam/client.go
Comment (reinkrul) [inline auth/api/iam/client.go:104]: unsupported scope might deserve its own error flow path?

File: docs/_static/auth/iam.yaml
Comment (reinkrul) [inline docs/_static/auth/iam.yaml:126]: Maybe reference RFC021?

File: docs/_static/auth/iam.yaml
Comment (reinkrul) [inline docs/_static/auth/iam.yaml:133]:
```suggestion
        a presentation can created together with a presentation submission.
```

File: docs/_static/auth/iam.yaml
Comment (reinkrul) [inline docs/_static/auth/iam.yaml:376]: add link to spec?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: auth/api/iam/api.go
Comment (woutslakhorst) [reply auth/api/iam/api.go:263]: The API gets them as slice.

File: auth/api/iam/client.go
Comment (woutslakhorst) [reply auth/api/iam/client.go:88]: You get it as a string from the metadata, so it would move the parsing to the caller. That would duplicate parsing code for every call.

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: auth/api/iam/api.go
Comment (woutslakhorst) [reply auth/api/iam/api.go]: Compile error due to missing methods on that type (oapi-codegen specific)

File: docs/_static/auth/iam.yaml
Comment (woutslakhorst) [reply docs/_static/auth/iam.yaml:152]: Removed specifics, they should be specified by use-cases

**Comment by woutslakhorst** [comment]
The PE suggestion won't work since you don't know which one of the backends you pick will actually give you access. The patient specific scope will only be checked after a client submitted a VP. If a client picks the wrong one, it'll fail and does not retry. There's no complexity on the CIS side since it's all handled by the node.

File: auth/api/iam/client.go
Comment (woutslakhorst) [reply auth/api/iam/client.go:88]: Putting ur.URL in the metadata will require custom marshalling/unmarshalling for the metadata struct.

File: auth/api/iam/api.go
Comment (woutslakhorst) [reply auth/api/iam/api.go:247]: backends do not have to check this, they only need to check if given VCs are enough for the given scope.

**Comment by woutslakhorst** [comment]
#2540 describes additional features required for Submission Requirement Feature.

File: auth/api/iam/api.go
Comment (gerardsn) [reply auth/api/iam/api.go:247]: I just realized that the PD.id generated by the backends will be thrown away when all PDs are merged

**Review by gerardsn — CHANGES_REQUESTED** [review]
just the one comment

File: auth/api/iam/client.go
Comment (woutslakhorst) [reply auth/api/iam/client.go]: I also changed the parsing so it will allow IPs in non-strictmode

File: auth/api/iam/client.go
Comment (gerardsn) [reply auth/api/iam/client.go:88]: Merging #2560 should make this easier

**Review by gerardsn — APPROVED** [review]
I don't think allowing IP in non-strictmode is going to be possible since parsing of did:web fails for IP addresses. But it might make unit tests easier if this is allowed?

**Comment by woutslakhorst** [comment]
@reinkrul waiting for your review

File: auth/api/iam/client.go
Comment (woutslakhorst) [reply auth/api/iam/client.go:88]: Hmmmm #2560 flipped the 2nd argument, although the PR build succeeds, after merging it'll fail....

File: auth/api/iam/error.go
Comment (reinkrul) [inline auth/api/iam/error.go:130]: consts are on line 35

File: auth/api/iam/error.go
Comment (reinkrul) [inline auth/api/iam/error.go:132]: could just return `OAuth2Error`? You're already unmarshalling it
Then you also have the description and stuff

**Review by reinkrul — APPROVED** [review]
Small remarks

**Pull request merged**
